src/Common.Lib/systemstring.cpp

```diff
@@ -0,0 +1,241 @@
+    {
+        m_hresult = FromString(lpzwStr, *this);
+    }
+
+    HRESULT SystemString::ToString(_Inout_ tstring& str)
+    {
+        return Convert(c_str(), str);
+    }
+
+    HRESULT SystemString::ToString(_Inout_ string& str)
+    {
+        str = *this;
+        return S_OK;
+    }
+
+    HRESULT SystemString::FromString(_In_ const WCHAR* lpzwStr, _Inout_ SystemString& result)
+    {
+        if (lpzwStr == nullptr)
+        {
+            result = "";
+            result.m_hresult = E_INVALIDARG;
+            return result.m_hresult;
+        }
+
+        size_t i = 0;
+        const WCHAR* p = lpzwStr;
+        // scan to a resonable length for the end of the string.
+        for (i = 0; i < MAX_STRING_LEN; ++i, ++p)
+        {
+            if (*p == (WCHAR)0)
+            {
+                break;
+            }
+        }
+
+        if (i > MAX_STRING_LEN)
+        {
+            result = "";
+            return E_FAIL;
+        }
+
+
+        Ciconv icnv(iconv_open("UTF-8", "UTF-16LE"));
+
+        // include the null terminator in the input string.
+        size_t inbytes = (i+1)*2;
+        size_t outbytes = (i+1)*4;
+        unique_ptr<char[]> buffer(new char[outbytes]);
+        char* input = (char*)lpzwStr;
+        char* output = buffer.get();
+        size_t count = iconv(icnv.m_iconv, &input, &inbytes, &output, &outbytes);
+        if (count == (size_t)-1)
+        {
+            HRESULT hr = ErrnoToHResult(errno);
+            result = "";
+            result.m_hresult = hr;
+            return result.m_hresult;
+        }
+
+        result = buffer.get();
+        return result.m_hresult;
+    }
+
+    HRESULT SystemString::FromString(_In_ const CHAR* lpzStr, _Inout_ SystemString& result)
+    {
+        if (lpzStr == nullptr)
+        {
+            result = "";
+            result.m_hresult = E_INVALIDARG;
+        }
+        else
+        {
+            result = SystemString(lpzStr);
+        }
+
+        return result.m_hresult;
+    }
+
+    HRESULT SystemString::Convert(_In_ const CHAR* lpzStr, _Inout_ tstring& result)
+    {
+        Ciconv icnv(iconv_open("UTF-16LE", "UTF-8"));
+
+        // Include the null terminator.
+        size_t inputLength = strnlen(lpzStr, MAX_STRING_LEN)  + 1;
+
+        if (inputLength >= MAX_STRING_LEN)
+        {
+            return E_BOUNDS;
+        }
+
+        size_t outputMax = (inputLength + 1) * 4;
+        unique_ptr<WCHAR[]> buffer(new WCHAR[outputMax]);
+
+        const char* pOrig = lpzStr;
+        char* pNew = (char*)buffer.get();
+        size_t count = iconv(icnv.m_iconv, (char**)&pOrig, &inputLength, &pNew, &outputMax);
+        if (count == (size_t)-1)
+        {
+            HRESULT hr = ErrnoToHResult(errno);
+            result = u"";
+            return hr;
+        }
+
+        result = buffer.get();
+        return S_OK;
+    }
+#else
+
+    SystemString::SystemString(_In_ const string& original)
+    {
+        m_hresult = FromString(original.c_str(), *this);
```

Comment:
>m_hresult = FromString(original.c_str(), *this) [](http://example.com/codeflow?start=8&length=47)

It would probably make more sense to reverse the order of these. That is, have FromString call the argumented constructor.